std/experimental/typecons.d

```diff
@@ -0,0 +1,731 @@
+                    staticMap!(generateFun, staticIota!(0, TargetMembers.length)));
+            }
+        }
+    }
+}
+
+// Internal class to support dynamic cross-casting
+private interface Structural(T)
+{
+    inout(T) _wrap_getSource() inout @safe pure nothrow;
+}
+
+private string unwrapExceptionText(Source, Target)()
+{
+    return Target.stringof~ " not wrapped into "~ Source.stringof;
+}
+
+/**
+ * Extract object which wrapped by $(D wrap).
+ *
+ * Params:
+ *     Target = Type to unwrap $(D src) to.
+ *     src = Item originally wrapped with a call to $(D wrap).
+ *
+ * Returns: The $(D Target) type from the wrapped type.
+ * If the type was not wrapped and is a class $(D null) is returned.
+ *
+ * Throws: ConvException when attempting to unwrap a struct
+ * which is not the target type.
+ */
+template unwrap(Target)
+{
+    static if (!isMutable!Target) {
+        alias unwrap = .unwrap!(Unqual!Target);
+    } else {
+        // strict downcast
+        auto unwrap(Source)(inout Source src) @trusted pure nothrow
+        if (is(Target : Source))
+        {
+            alias T = Select!(is(Source == shared), shared Target, Target);
+            return dynamicCast!(inout T)(src);
+        }
+        // structural downcast for struct target
+        auto unwrap(Source)(inout Source src)
+        if (is(Target == struct))
+        {
+            alias T = Select!(is(Source == shared), shared Target, Target);
+            Object upCastSource = dynamicCast!(Object)(src);   // remove qualifier
+            do
+            {
+                if (auto a = dynamicCast!(Structural!Object)(upCastSource))
+                {
+                    upCastSource = a._wrap_getSource();
+                }
+                else if (auto a = dynamicCast!(Structural!T)(upCastSource))
+                {
+                    return a._wrap_getSource();
+                }
+                else
+                {
+                    static if (hasMember!(Source, "_wrap_getSource"))
+                        return unwrap!Target(src._wrap_getSource());
+                    else
+                        break;
+                }
+            } while (upCastSource);
+            import std.conv: ConvException;
+            throw new ConvException(unwrapExceptionText!(Source,Target));
+        }
+        // structural downcast for class target
+        auto unwrap(Source)(inout Source src)
+        if (!is(Target : Source) && !is(Target == struct))
+        {
+            alias T = Select!(is(Source == shared), shared Target, Target);
+            Object upCastSource = dynamicCast!(Object)(src);   // remove qualifier
+            do
+            {
+                if (auto a = dynamicCast!(Structural!Object)(upCastSource))
+                {
+                    if (auto d = dynamicCast!(inout T)(upCastSource = a._wrap_getSource()))
+                        return d;
+                }
+                else if (auto a = dynamicCast!(Structural!T)(upCastSource))
```

Comment:
That is not the case, I've added comments on the purpose of each condition in hopes that it will be clear.